File: swift/Sources/benchmarks/JsonParseDom.swift
import Foundation

final class JsonParseDom: BenchmarkProtocol {
  private var text: String = ""
  private var resultVal: UInt32 = 0

  func prepare() {
    let generator = JsonGenerate()
    generator.n = configValue("coords") ?? 0
    generator.prepare()
    generator.run(iterationId: 0)

    let mirror = Mirror(reflecting: generator)
    for child in mirror.children {
      if child.label == "text" {
        text = child.value as? String ?? ""
        break
      }
    }
  }

  private func calcDom(_ text: String) -> (Double, Double, Double) {
    guard let data = text.data(using: .utf8) else {
      return (0, 0, 0)
    }
    do {
      let json = try JSONSerialization.jsonObject(with: data)
      guard let dict = json as? [String: Any],
        let coordinates = dict["coordinates"] as? [[String: Any]]
      else {
        return (0, 0, 0)
      }
      var x = 0.0
      var y = 0.0
      var z = 0.0

      for coord in coordinates {
        let coordX = coord["x"] as? Double ?? 0
        let coordY = coord["y"] as? Double ?? 0
        let coordZ = coord["z"] as? Double ?? 0
        x += coordX
        y += coordY
        z += coordZ
      }
      let len = Double(coordinates.count)
      return (x / len, y / len, z / len)
    } catch {
      return (0, 0, 0)
    }
  }

  func run(iterationId: Int) {
    let (x, y, z) = calcDom(text)
    resultVal &+= Helper.checksumF64(x) &+ Helper.checksumF64(y) &+ Helper.checksumF64(z)
  }

  var checksum: UInt32 {
    return resultVal
  }
}

File: nim/src/benchmarks/jsonparsedom.nim
import std/[math, json]
import ../benchmark
import ../helper
import jsonbench_common

type
  JsonParseDom* = ref object of Benchmark
    text: string
    resultVal: uint32

proc newJsonParseDom(): Benchmark =
  JsonParseDom()

method name(self: JsonParseDom): string = "JsonParseDom"

method prepare(self: JsonParseDom) =
  let n = self.config_val("coords")
  self.text = getJsonText(n)
  self.resultVal = 0

method run(self: JsonParseDom, iteration_id: int) =
  let parsed = parseJson(self.text)

  var xSum, ySum, zSum: float
  var len = 0

  let coordinates = parsed["coordinates"]

  for coordNode in coordinates:
    xSum += coordNode["x"].getFloat()
    ySum += coordNode["y"].getFloat()
    zSum += coordNode["z"].getFloat()
    inc len

  if len > 0:
    let x = xSum / float(len)
    let y = ySum / float(len)
    let z = zSum / float(len)

    self.resultVal = self.resultVal + checksumF64(x) + checksumF64(y) +
        checksumF64(z)

method checksum(self: JsonParseDom): uint32 =
  self.resultVal

registerBenchmark("JsonParseDom", newJsonParseDom)
